internal/pkg/composable/providers/host/host_test.go

```diff
@@ -21,8 +21,12 @@ import (
 )
 
 func TestContextProvider(t *testing.T) {
+	log, err := logger.New("host_test", false)
```

Comment:
Done in 022b8b43451d11e2d297f09a750eaa979a0fddd1.

Reply:
Hmmm, the test I wrote is failing in CI.  For some reason, the DNS mocking doesn't seem to be working here but it worked in https://github.com/elastic/beats/pull/34946 🤔

Reply:
I'm able to reproduce this failure locally.  Turns out, if I set my machine's hostname to something like "foo.bar.local", i.e. containing a domain, the mock DNS resolution doesn't work and the test fails.  But if I set my machine's hostname to something like "foo", the mock DNS resolution comes into play and the test succeeds.

I ran the Beats test locally too with the two different hostname configurations and had the same results as I mentioned above.

@cmacknz I'll spend some more time investigating what's going on under the hood, but I'm leaning towards not using this mock DNS resolution technique in our tests in case that makes them flaky.  Instead I think an end-to-end test where we can actually set the hostname of the machine might be more reliable.  WDYT?

Reply:
@ycombinator `.local` is a bit of a special domain (mDNS anyone?) so maybe the go-mockdns doesn't cover that particular usecase. Did you have the issue mocking a domain name also with non mDNS domains ?

Reply:
If this is isolated to some special behaviour related to `.local` then I vote we keep the mock DNS test. Even if we can't figure it out let's keep the test without `.local` and just add a comment about this special case. 

It could be a bug in the mock DNS package, and while it is annoying not to know why it fails it still provides value with this quirk.

Emulating the DNS behaviour in unit tests is really valuable, it is a much faster way to test this rather than relying on manual or even automated E2E testing.

Reply:
The behavior is, unfortunately, not isolated to a `.local` hostname:

```
$ hostname -f
Shaunaks-MBP.attlocal.net
$ go test github.com/elastic/beats/v7/libbeat/processors/add_host_metadata -test.run TestFQDNLookup -v -count 1
--- FAIL: TestFQDNLookup (0.01s)
    --- FAIL: TestFQDNLookup/lookup_succeeds (0.01s)
        add_host_metadata_test.go:540: os.Hostname(): Shaunaks-MBP.attlocal.net
        add_host_metadata_test.go:573:
                Error Trace:    /Users/shaunak/development/github/beats/libbeat/processors/add_host_metadata/add_host_metadata_test.go:573
                Error:          Not equal:
                                expected: 0
                                actual  : 1
                Test:           TestFQDNLookup/lookup_succeeds
    --- FAIL: TestFQDNLookup/lookup_fails (0.00s)
        add_host_metadata_test.go:540: os.Hostname(): Shaunaks-MBP.attlocal.net
        add_host_metadata_test.go:573:
                Error Trace:    /Users/shaunak/development/github/beats/libbeat/processors/add_host_metadata/add_host_metadata_test.go:573
                Error:          Not equal:
                                expected: 1
                                actual  : 2
                Test:           TestFQDNLookup/lookup_fails
FAIL
FAIL    github.com/elastic/beats/v7/libbeat/processors/add_host_metadata        0.720s
FAIL
```

Whereas, the same tests with a short domain name (no domain component), pass.  Also note that, in this case, there's output from the DNS mock, which tells me that it's likely not coming into play in the dotted hostname scenario above.

```
$ sudo scutil --set HostName shortypants
$ hostname -f
shortypants
$ go test github.com/elastic/beats/v7/libbeat/processors/add_host_metadata -test.run TestFQDNLookup -v -count 1
=== RUN   TestFQDNLookup
=== RUN   TestFQDNLookup/lookup_succeeds
    add_host_metadata_test.go:540: os.Hostname(): shortypants
mockdns server: 2023/03/31 12:40:20 DNS TRACE ;; opcode: QUERY, status: NOERROR, id: 43628
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;shortypants.   IN       AAAA

;; ANSWER SECTION:
shortypants.    9999    IN      CNAME   foo.bar.baz.
mockdns server: 2023/03/31 12:40:20 DNS TRACE ;; opcode: QUERY, status: NOERROR, id: 34242
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;shortypants.   IN       A

;; ANSWER SECTION:
shortypants.    9999    IN      CNAME   foo.bar.baz.
foo.bar.baz.    9999    IN      A       1.1.1.1
=== RUN   TestFQDNLookup/lookup_fails
    add_host_metadata_test.go:540: os.Hostname(): shortypants
mockdns server: 2023/03/31 12:40:20 DNS TRACE ;; opcode: QUERY, status: NOERROR, id: 54098
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;shortypants.   IN       A
mockdns server: 2023/03/31 12:40:20 DNS TRACE ;; opcode: QUERY, status: NOERROR, id: 46393
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;shortypants.   IN       AAAA
mockdns server: 2023/03/31 12:40:20 DNS TRACE ;; opcode: QUERY, status: NOERROR, id: 42738
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;shortypants.   IN       AAAA
mockdns server: 2023/03/31 12:40:20 DNS TRACE ;; opcode: QUERY, status: NOERROR, id: 27118
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;shortypants.   IN       A
--- PASS: TestFQDNLookup (0.01s)
    --- PASS: TestFQDNLookup/lookup_succeeds (0.00s)
    --- PASS: TestFQDNLookup/lookup_fails (0.00s)
PASS
ok      github.com/elastic/beats/v7/libbeat/processors/add_host_metadata        0.735s
```

Reply:
Hmm. If you think this isn't valuable then feel free to remove it in favor of using an E2E test for this. You are the one who has to maintain this, do whatever you think makes sense here.

Reply:
I think it's valuable to have a unit test but I'm not able to get the mock DNS resolver to always be invoked by this test. It looks like `srv.PatchNet(net.DefaultResolver)` will override the default resolver's `Dial` method so it uses the mock DNS server for resolution.  But something in `net.LookupCNAME` must be short-circuiting and not even getting around to calling this `Dial` method. 

The unit test is currently failing in CI even with a short hostname: https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp/detail/PR-2218/108/pipeline#step-673-log-232

I've looked through and played with the https://github.com/foxcpp/go-mockdns API but I'm not seeing anything that'll help.  At this point, I think it will be better to write an E2E test for this where we can set the machine's hostname to something we know will resolve and then to something we know will not resolve.